File: backend/src/main/java/br/ufpr/sept/so2/shared/api/GlobalExceptionHandler.java
package br.ufpr.sept.so2.shared.api;

import br.ufpr.sept.so2.shared.domain.exception.ConflitoEstadoException;
import br.ufpr.sept.so2.shared.domain.exception.DadoInvalidoException;
import br.ufpr.sept.so2.shared.domain.exception.RecursoNaoEncontradoException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.security.access.AccessDeniedException;
import org.springframework.security.core.AuthenticationException;
import org.springframework.validation.FieldError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestControllerAdvice;
import org.springframework.web.context.request.WebRequest;

import java.net.URI;
import java.time.OffsetDateTime;
import java.util.List;
import java.util.Map;
import java.util.UUID;

@RestControllerAdvice
public class GlobalExceptionHandler {

    private static final Logger LOG = LoggerFactory.getLogger(GlobalExceptionHandler.class);
    private static final String ERROR_BASE = "https://secretariaonline.ufpr.br/errors/";

    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ProblemDetail handleValidation(MethodArgumentNotValidException ex) {
        List<Map<String, String>> errors = ex.getBindingResult().getAllErrors().stream()
                .map(error -> {
                    if (error instanceof FieldError fieldError) {
                        return Map.of(
                                "campo", fieldError.getField(),
                                "mensagem", fieldError.getDefaultMessage() == null ? "inválido" : fieldError.getDefaultMessage()
                        );
                    }
                    return Map.of("mensagem", error.getDefaultMessage() == null ? "inválido" : error.getDefaultMessage());
                })
                .toList();
        ProblemDetail detail = problemDetail(
                HttpStatus.UNPROCESSABLE_ENTITY,
                "Dados inválidos",
                "A requisição contém dados inválidos. Verifique os campos.",
                "validation-error"
        );
        detail.setProperty("erros", errors);
        return detail;
    }

    @ExceptionHandler(RecursoNaoEncontradoException.class)
    public ProblemDetail handleNotFound(RecursoNaoEncontradoException ex) {
        return problemDetail(HttpStatus.NOT_FOUND, "Recurso não encontrado", ex.getMessage(), "not-found");
    }

    @ExceptionHandler(DadoInvalidoException.class)
    public ProblemDetail handleInvalid(DadoInvalidoException ex) {
        return problemDetail(HttpStatus.UNPROCESSABLE_ENTITY, "Dados inválidos", ex.getMessage(), "validation-error");
    }

    @ExceptionHandler(ConflitoEstadoException.class)
    public ProblemDetail handleConflict(ConflitoEstadoException ex) {
        return problemDetail(HttpStatus.CONFLICT, "Conflito de estado", ex.getMessage(), "conflict");
    }

    @ExceptionHandler(IllegalArgumentException.class)
    public ProblemDetail handleIllegalArgument(IllegalArgumentException ex) {
        return problemDetail(HttpStatus.BAD_REQUEST, "Dados inválidos", ex.getMessage(), "bad-request");
    }

    @ExceptionHandler(AccessDeniedException.class)
    public ProblemDetail handleAccessDenied(AccessDeniedException ex, WebRequest request) {
        LOG.warn("Acesso negado: {} - {}", request.getDescription(false), ex.getMessage());
        return problemDetail(HttpStatus.FORBIDDEN, "Acesso negado", "Você não tem permissão para esta operação.", "access-denied");
    }

    @ExceptionHandler(AuthenticationException.class)
    public ProblemDetail handleAuthentication(AuthenticationException ex) {
        return problemDetail(HttpStatus.UNAUTHORIZED, "Não autenticado", "Token inválido ou expirado.", "authentication-required");
    }

    @ExceptionHandler(Exception.class)
    public ProblemDetail handleGeneric(Exception ex, WebRequest request) {
        String incidentId = "INC-" + OffsetDateTime.now().getYear() + "-"
                + UUID.randomUUID().toString().replace("-", "").substring(0, 4);
        LOG.error("Erro inesperado [{}] em {}: {}", incidentId, request.getDescription(false), ex.getMessage(), ex);
        ProblemDetail detail = problemDetail(
                HttpStatus.INTERNAL_SERVER_ERROR,
                "Erro interno",
                "Ocorreu um erro inesperado. Tente novamente ou contate o suporte.",
                "internal-error"
        );
        detail.setProperty("incidentId", incidentId);
        return detail;
    }

    private ProblemDetail problemDetail(HttpStatus status, String title, String detail, String type) {
        ProblemDetail problem = ProblemDetail.forStatusAndDetail(status, detail);
        problem.setTitle(title);
        problem.setType(URI.create(ERROR_BASE + type));
        problem.setProperty("timestamp", OffsetDateTime.now().toString());
        return problem;
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/shared/api/GlobalExceptionHandler.kt
package br.ufpr.sept.so2.shared.api

import br.ufpr.sept.so2.modules.iam.domain.SenhaReutilizadaException
import br.ufpr.sept.so2.shared.domain.exception.AcessoNegadoException
import br.ufpr.sept.so2.shared.domain.exception.ConflitoEstadoException
import br.ufpr.sept.so2.shared.domain.exception.CredenciaisInvalidasException
import br.ufpr.sept.so2.shared.domain.exception.DadoInvalidoException
import br.ufpr.sept.so2.shared.domain.exception.RateLimitExcedidoException
import br.ufpr.sept.so2.shared.domain.exception.RecursoNaoEncontradoException
import br.ufpr.sept.so2.shared.domain.exception.TokenResetInvalidoException
import org.slf4j.LoggerFactory
import org.springframework.http.HttpStatus
import org.springframework.http.ProblemDetail
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.security.access.AccessDeniedException
import org.springframework.security.core.AuthenticationException
import org.springframework.validation.FieldError
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice
import org.springframework.web.context.request.WebRequest
import java.net.URI
import java.time.OffsetDateTime
import java.util.UUID

@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(HttpMessageNotReadableException::class)
    fun handleUnreadable(ex: HttpMessageNotReadableException): ProblemDetail =
        problemDetail(
            HttpStatus.BAD_REQUEST,
            "Dados inválidos",
            "Corpo da requisição ausente ou inválido.",
            "bad-request",
        )

    @ExceptionHandler(MethodArgumentNotValidException::class)
    fun handleValidation(ex: MethodArgumentNotValidException): ProblemDetail {
        val errors = ex.bindingResult.allErrors.map { error ->
            if (error is FieldError) {
                mapOf(
                    "campo" to error.field,
                    "mensagem" to (error.defaultMessage ?: "inválido"),
                )
            } else {
                mapOf("mensagem" to (error.defaultMessage ?: "inválido"))
            }
        }
        val detail = problemDetail(
            HttpStatus.UNPROCESSABLE_ENTITY,
            "Dados inválidos",
            "A requisição contém dados inválidos. Verifique os campos.",
            "validation-error",
        )
        detail.setProperty("erros", errors)
        return detail
    }

    @ExceptionHandler(RecursoNaoEncontradoException::class)
    fun handleNotFound(ex: RecursoNaoEncontradoException): ProblemDetail =
        problemDetail(HttpStatus.NOT_FOUND, "Recurso não encontrado", ex.message, "not-found")

    @ExceptionHandler(SenhaReutilizadaException::class)
    fun handleSenhaReutilizada(ex: SenhaReutilizadaException): ProblemDetail =
        problemDetail(HttpStatus.UNPROCESSABLE_ENTITY, "Senha reutilizada", ex.message, "senha-reutilizada")

    @ExceptionHandler(DadoInvalidoException::class)
    fun handleInvalid(ex: DadoInvalidoException): ProblemDetail =
        problemDetail(HttpStatus.UNPROCESSABLE_ENTITY, "Dados inválidos", ex.message, "validation-error")

    @ExceptionHandler(CredenciaisInvalidasException::class)
    fun handleCredenciais(ex: CredenciaisInvalidasException): ProblemDetail =
        problemDetail(
            HttpStatus.UNAUTHORIZED,
            "Não autenticado",
            CredenciaisInvalidasException.MENSAGEM,
            "authentication-required",
        )

    @ExceptionHandler(TokenResetInvalidoException::class)
    fun handleResetToken(ex: TokenResetInvalidoException): ProblemDetail =
        problemDetail(
            HttpStatus.UNAUTHORIZED,
            "Não autenticado",
            TokenResetInvalidoException.MENSAGEM,
            "authentication-required",
        )

    @ExceptionHandler(RateLimitExcedidoException::class)
    fun handleRateLimit(ex: RateLimitExcedidoException): ProblemDetail {
        val detail = problemDetail(
            HttpStatus.TOO_MANY_REQUESTS,
            "Muitas tentativas",
            ex.message,
            "rate-limit",
        )
        detail.setProperty("retryAfterSeconds", ex.retryAfterSeconds)
        return detail
    }

    @ExceptionHandler(AcessoNegadoException::class)
    fun handleAcessoNegado(ex: AcessoNegadoException): ProblemDetail =
        problemDetail(HttpStatus.FORBIDDEN, "Acesso negado", ex.message, "access-denied")

    @ExceptionHandler(ConflitoEstadoException::class)
    fun handleConflict(ex: ConflitoEstadoException): ProblemDetail =
        problemDetail(HttpStatus.CONFLICT, "Conflito de estado", ex.message, "conflict")

    @ExceptionHandler(IllegalArgumentException::class)
    fun handleIllegalArgument(ex: IllegalArgumentException): ProblemDetail =
        problemDetail(HttpStatus.BAD_REQUEST, "Dados inválidos", ex.message, "bad-request")

    @ExceptionHandler(AccessDeniedException::class)
    fun handleAccessDenied(ex: AccessDeniedException, request: WebRequest): ProblemDetail {
        LOG.warn("Acesso negado: {} - {}", request.getDescription(false), ex.message)
        return problemDetail(
            HttpStatus.FORBIDDEN,
            "Acesso negado",
            "Você não tem permissão para esta operação.",
            "access-denied",
        )
    }

    @ExceptionHandler(AuthenticationException::class)
    fun handleAuthentication(ex: AuthenticationException): ProblemDetail =
        problemDetail(
            HttpStatus.UNAUTHORIZED,
            "Não autenticado",
            "Token inválido ou expirado.",
            "authentication-required",
        )

    @ExceptionHandler(Exception::class)
    fun handleGeneric(ex: Exception, request: WebRequest): ProblemDetail {
        val incidentId = "INC-" + OffsetDateTime.now().year + "-" +
            UUID.randomUUID().toString().replace("-", "").substring(0, 4)
        LOG.error("Erro inesperado [{}] em {}: {}", incidentId, request.getDescription(false), ex.message, ex)
        val detail = problemDetail(
            HttpStatus.INTERNAL_SERVER_ERROR,
            "Erro interno",
            "Ocorreu um erro inesperado. Tente novamente ou contate o suporte.",
            "internal-error",
        )
        detail.setProperty("incidentId", incidentId)
        return detail
    }

    private fun problemDetail(status: HttpStatus, title: String, detail: String?, type: String): ProblemDetail {
        val problem = ProblemDetail.forStatusAndDetail(status, detail)
        problem.title = title
        problem.type = URI.create(ERROR_BASE + type)
        problem.setProperty("timestamp", OffsetDateTime.now().toString())
        return problem
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(GlobalExceptionHandler::class.java)
        private const val ERROR_BASE = "https://secretariaonline.ufpr.br/errors/"
    }
}
